File: Solutions/Python/3sum.py
"""
    [MEDIUM]
    15. 3Sum

    Concepts:
    - two pointers
    - sorting

    Stats:
        Runtime | 522 ms    [Beats 77.51%]
        Memory  | 19.14 MB  [Beats 99.28%]
"""

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        """
            Use two pointers, one at each end of sorted array
        """
        triplets = set()
        nums.sort()
        
        n = len(nums)

        for i in range(n):
            # skip duplicate values
            if i > 0 and nums[i] == nums[i-1]:
                continue
            
            # set target to be found using other two indices
            target = -nums[i]

            j = i + 1
            k = n - 1

            while j < k:
                val = nums[j] + nums[k]

                if val == target:
                    triplets.add(tuple([nums[v] for v in [i, j, k]]))
                    j += 1
                    k -= 1
                elif val > target:
                    k -= 1
                else:
                    j += 1

        return list(triplets)

File: Solutions/Cpp/3sum.cpp
/*
    [MEDIUM]
    15. 3Sum

    Concepts:
    - two pointers
    - sorting

    Stats:
        Runtime | 60 ms     [Beats 30.95%]
        Memory  | 35.32 MB  [Beats 20.82%]
*/

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        std::set<std::vector<int>> triplets = {};
        int n = nums.size();
        int target = 0;
        int j, k;

        std::sort(nums.begin(), nums.end());

        for (int i = 0; i < n; i++) {
            // skip duplicate values
            if (i && nums[i] == nums[i-1])
                continue;

            // set target to be found using other two indices
            target = -nums[i];

            j = i + 1;
            k = n - 1;

            while (j < k) {
                if (nums[j] + nums[k] == target) {
                    triplets.insert({nums[i], nums[j], nums[k]});
                    j++;
                    k--;
                } else if (nums[j] + nums[k] > target) {
                    k--;
                } else {
                    j++;
                }
            }
        }
        
        return std::vector<std::vector<int>>{triplets.begin(), triplets.end()};
    }
};